util/rowcodec/decoder.go

```diff
@@ -183,31 +295,158 @@ func (decoder *Decoder) decodeColData(colIdx int, colData []byte, chk *chunk.Chu
 	case mysql.TypeDuration:
 		var dur types.Duration
 		dur.Duration = time.Duration(decodeInt(colData))
-		dur.Fsp = int8(ft.Decimal)
+		dur.Fsp = int8(col.Decimal)
 		chk.AppendDuration(colIdx, dur)
 	case mysql.TypeEnum:
 		// ignore error deliberately, to read empty enum value.
-		enum, err := types.ParseEnumValue(ft.Elems, decodeUint(colData))
+		enum, err := types.ParseEnumValue(col.Elems, decodeUint(colData))
 		if err != nil {
 			enum = types.Enum{}
 		}
 		chk.AppendEnum(colIdx, enum)
 	case mysql.TypeSet:
-		set, err := types.ParseSetValue(ft.Elems, decodeUint(colData))
+		set, err := types.ParseSetValue(col.Elems, decodeUint(colData))
 		if err != nil {
 			return err
 		}
 		chk.AppendSet(colIdx, set)
 	case mysql.TypeBit:
-		byteSize := (ft.Flen + 7) >> 3
+		byteSize := (col.Flen + 7) >> 3
 		chk.AppendBytes(colIdx, types.NewBinaryLiteralFromUint(decodeUint(colData), byteSize))
 	case mysql.TypeJSON:
 		var j json.BinaryJSON
 		j.TypeCode = colData[0]
 		j.Value = colData[1:]
 		chk.AppendJSON(colIdx, j)
 	default:
-		return errors.Errorf("unknown type %d", ft.Tp)
+		return errors.Errorf("unknown type %d", col.Tp)
 	}
 	return nil
 }
+
+// BytesDecoder decodes the row to old datums bytes.
+type BytesDecoder struct {
+	decoder
+	defBytes []func() ([]byte, error)
```

Comment:
If we make the `defBytes` accept an `i` argument, then we only need to implement one function.